### PR TITLE
Use sorted index based filtering only for dictionary encoded column

### PR DESCRIPTION
**Pull request by siddharthteotia**
Currently we build sorted index only if the column is dictionary encoded. However, when we write `isSorted` in on-disk segment metadata, we write on the basis of pre-index stats collector. So, for a sorted column without dictionary, segment metadata will still indicate column as sorted

`properties.setProperty(getKeyFor(column, IS_SORTED), String.valueOf(columnIndexCreationInfo.isSorted()));`

During query processing, when we create filter operator, we check the data source metadata to see if the column is sorted and create sorted index based filter operator. However, using this operator for any sorted raw column will lead to the following error stack since we end up using a raw value based predicate evaluator for a dictionary based filter operator. 

The solution is to do the additional check on data source to see if the column is dictionary encoded or not

````
java.lang.UnsupportedOperationException
181762         at org.apache.pinot.core.operator.filter.predicate.BaseRawValueBasedPredicateEvaluator.getMatchingDictIds(BaseRawValueBasedPredicateEvaluator.java:40)
181763         at org.apache.pinot.core.operator.filter.SortedIndexBasedFilterOperator.getNextBlock(SortedIndexBasedFilterOperator.java:68)
181764         at org.apache.pinot.core.operator.filter.SortedIndexBasedFilterOperator.getNextBlock(SortedIndexBasedFilterOperator.java:35)
181765         at org.apache.pinot.core.operator.BaseOperator.nextBlock(BaseOperator.java:49)
181766         at org.apache.pinot.core.operator.DocIdSetOperator.getNextBlock(DocIdSetOperator.java:62)
181767         at org.apache.pinot.core.operator.DocIdSetOperator.getNextBlock(DocIdSetOperator.java:35)
181768         at org.apache.pinot.core.operator.BaseOperator.nextBlock(BaseOperator.java:49)
181769         at org.apache.pinot.core.operator.ProjectionOperator.getNextBlock(ProjectionOperator.java:57)
181770         at org.apache.pinot.core.operator.ProjectionOperator.getNextBlock(ProjectionOperator.java:30)`
```

### CODE REVIEW

**Review by mcvsubbu — APPROVED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java [hunk 1]
```diff
@@ -47,9 +47,15 @@ public static BaseFilterOperator getLeafFilterOperator(PredicateEvaluator predic
...
+    // Currently sorted index based filtering is supported only for
+    // dictionary encoded columns. The on-disk segment metadata
+    // will indicate if the column is sorted or not regardless of
+    // whether it is raw or dictionary encoded. Here when creating
+    // the filter operator, we need to make sure that sort filter
+    // operator is used only if the column is sorted and has dictionary.
     Predicate.Type predicateType = predicateEvaluator.getPredicateType();
     if (predicateType == Predicate.Type.RANGE) {
-      if (dataSource.getDataSourceMetadata().isSorted()) {
+      if (dataSource.getDataSourceMetadata().isSorted() && (dataSource.getDictionary() != null)) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java:58]: (nit)
```suggestion
      if (dataSource.getDataSourceMetadata().isSorted() && dataSource.getDictionary() != null) {
```

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java [hunk 2]
```diff
@@ -59,7 +65,7 @@ public static BaseFilterOperator getLeafFilterOperator(PredicateEvaluator predic
     } else if (predicateType == Predicate.Type.REGEXP_LIKE) {
       return new ScanBasedFilterOperator(predicateEvaluator, dataSource, numDocs);
     } else {
-      if (dataSource.getDataSourceMetadata().isSorted()) {
+      if (dataSource.getDataSourceMetadata().isSorted() && (dataSource.getDictionary() != null)) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java:68]: (nit)
```suggestion
      if (dataSource.getDataSourceMetadata().isSorted() && dataSource.getDictionary() != null) {
```

**Review by Jackie-Jiang — APPROVED** [review]
(Optional) IMO it is cleaner if we check the dictionary when initializing the data source metadata since we don't support raw sorted index. In `ImmutableDataSource.java`, you can change the constructor of `ImmutableDataSourceMetadata` to `_sorted = columnMetadata.isSorted() && columnMetadata.hasDictionary()`. When we support raw sorted index in the future, we can change it back. Wdyt?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java (hunk 1 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java (hunk 2 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java]: done

**Pull request merged**